west build enhancements

mbolivar opened the pull request:
This implements some of the west build enhancements requested the day after the TSC F2F and discussed a bit since then:

- Fixes https://github.com/zephyrproject-rtos/west/issues/252 (via build.generator config option, not a command line option)
- Fixes https://github.com/zephyrproject-rtos/west/issues/253
- Fixes https://github.com/zephyrproject-rtos/west/issues/254
- Fixes https://github.com/zephyrproject-rtos/west/issues/255
- Fixes https://github.com/zephyrproject-rtos/west/issues/256

This also implements a `build.board` config option to set a default board (which is used as a fallback after CACHED_BOARD in the CMake cache, BOARD in the environment, and --board on the command line), which I think is a better idea than what's proposed in https://github.com/zephyrproject-rtos/west/issues/251.

mbolivar-nordic committed:
west: build: use CMAKE_HOME_DIRECTORY to locate source

Rather than relying on APPLICATION_SOURCE_DIR to be in the cache, just
use CMAKE_HOME_DIRECTORY instead.

https://cmake.org/cmake/help/latest/variable/CMAKE_HOME_DIRECTORY.html

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: re-word and extend west build documentation

Try to make the language simpler. Introduce each example with a
paragraph that says "To <DO XYZ>" to make them easier to find.

Move the config options to the documentation for that command to make
them easier to find.

Add some more examples for use cases we've gotten questions about.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: add --cmake-only to west build

This forces CMake to run, but skips running any build system target
afterwards.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: add build.generator config option for west build

This can be used to override the default CMake generator
permanently. Its values are the same as those acceptable to cmake's -G
option.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: fix up west build verbosity levels

Adjust them so "west build -v" prints ZEPHYR_BASE and any CMake
commands, but none of the other more esoteric bits of information.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: add --dry-run flags to west build

Analogously to the Make options with the same names, these print the
commands which would have run without running them.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -253,13 +253,13 @@ def _setup_source_dir(self):
             source_dir = self.args.source_dir
         elif self.cmake_cache:
-            source_dir = self.cmake_cache.get('APPLICATION_SOURCE_DIR')
+            source_dir = self.cmake_cache.get('CMAKE_HOME_DIRECTORY')
```
Should we use `CMAKE_SOURCE_DIR` instead?
And is this still valid when compiling an out-of-tree app?

SebastianBoe replied:
We should, but unfortunately it is not in the cache, so we must use the internal variable instead.

carlescufi commented on doc/guides/west/config.rst:
```diff
@@ -164,6 +165,13 @@ extension commands (found in :file:`scripts/west_commands`).
    * - Option
      - Description
+   * - ``build.board``
+     - String. If given, this the board used by :ref:`west build
```
Nice to have a default board! I remember thinking about this as well a while back.

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -58,6 +62,12 @@
   cmake_opt             Extra options to pass to CMake; implies -c
 '''
 
+def config_get(option, fallback):
```
Shouldn't this be used for line 164:
`pristine = config.get('build', 'pristine', fallback='never')`

mbolivar replied:
Sure, I can convert all the existing users too.

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -175,26 +183,27 @@ def do_run(self, args, remainder):
+        board, origin = None, None
+        config_board = config_get('board', None)
+        if self.cmake_cache:
+            board, origin = (self.cmake_cache.get('CACHED_BOARD'),
```
Can this ever return `None` now that I think of it?

mbolivar replied:
Not if the build directory is valid

mbolivar replied:
Was on mobile before. To amplify: if the build directory is messed up somehow, board comes back None, but `_run_cmake` complains asking for `--board`. At that point, if the user passes `--board` and the build *still* doesn't work, something is quite wrong with the build directory and they probably just need to delete it and start over.

carlescufi commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -53,86 +53,124 @@ Building: ``west build``
+build directory::
 
-  west build -t clean
+  west build -t pristine
```
We should use `-t run` here. There's already `-p` for `pristine` so not the greatest example

mbolivar replied:
I'm switching to this on purpose to make the -p switch more clear

mbolivar replied:
I've added several different target examples now.

carlescufi commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -53,86 +53,124 @@ Building: ``west build``
   west build -c
 
+Configuration Options
+=====================
```
Why move this here? I thought a centralized source for all configuration options was a good idea actually. It gave a clear overview of all the ways you can configure west in. Maybe create a second table for extension commands in `config.rst` and link to it from here?

mbolivar replied:
For one thing, the other page is about "core" west, and this is an extension. For another, as I said in the commit message, i think this will be a lot easier for people to find.

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -137,6 +137,8 @@ def do_add_parser(self, parser_adder):
                             application.''')
         parser.add_argument('-c', '--cmake', action='store_true',
                             help='Force CMake to run')
+        parser.add_argument('--just-cmake', action='store_true',
```
I would prefer `--cmake-only`. I think that's the usual way of restricting the scope of the execution of a command-line tool, although I might be biased by the ones I use.

mbolivar replied:
Sure, np

carlescufi commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -157,6 +159,10 @@ You can :ref:`configure <west-config-cmd>` ``west build`` using these options.
      - Boolean, default ``true``. If ``false``, disables warnings when
        ``west build`` can't figure out the target board.
+   * - ``build.generator``
+     - String, default ``Ninja``. The `CMake Generator`_ to use to create a
```
Do you need quotes in `.westconfig`? Because some of the genertor names have spaces in them like "Unix Makefiles".

mbolivar replied:
Values with spaces are natively supported by the Python config library.

mbolivar replied:
Ah, but there is a bug; it should be config_get('generator', ...), Will fix, thanks!

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -139,6 +139,10 @@ def do_add_parser(self, parser_adder):
                             help='Force CMake to run')
         parser.add_argument('--just-cmake', action='store_true',
                             help="Just run CMake; don't build. Implies -c.")
+        parser.add_argument('-n', '--just-print', '--dry-run', '--recon',
```
do we really need that many names? `-n`/`--dry-run` should be enough?

mbolivar replied:
Those are the names Make uses. I'm making sure people get all of them.

carlescufi commented on scripts/west_commands/cmake.py:
```diff
@@ -39,6 +49,11 @@ def run_cmake(args, cwd=None, capture_output=False):
         kwargs['cwd'] = cwd
+
+    if dry_run:
+        log.inf('Dry run:', quote_sh_list(cmd))
```
`log.inf` is always printed right?

mbolivar replied:
For now, yeah. We may add a -q switch later, but we don't have one yet.

carlescufi commented:
Thanks for the series! Great stuff, just minor suggestions and nits.

mbolivar commented:
> This can be dismissed as out-of-scope, but it would be very helpful if these two classes could be placed somewhere that makes it clean and easy for scripts in zephyr/scripts to include it.

Yeah, I agree, but this requires refactoring `west.log` usage a bit. It's on my TODO list; I'm thinking of just switching to the standard Python `logging` module and setting up handlers that produce equivalent output to `west.log` today, but I'm not convinced that will work, and there's backwards compatibility with LTS to consider. So if you're OK with leaving this to another PR, I'd appreciate it.

mbolivar-nordic committed:
scripts: add helpers to west build for getting config options

These will be used more in a later patch.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: improve west build's board handling

- Respect the BOARD environment setting.
- Don't require --force if the board can't be figured out: it might be
  set in CMakeLists.txt, for example. Instead, downgrade to a warning
  which can be disabled with "west config build.board_warn false".
- Add a build.board configuration option used as a final BOARD fallback
  after CACHED_BOARD (in the CMake cache), --board (command line), and
  BOARD (environment).
- Keep the config docs up to date.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: flake8 cleanup for west build

Clean up coding nits to match west's style.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: add --build-opt option to west build

This allows passing arguments to the underlying build tool (ninja,
make, etc.).

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: make west build -h fit on a screen

The west build --help output no longer fits in a single page.  Move
details and examples into the documentation, so the -h output doesn't
require scrolling around.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: fix west boards tip

Doc fix for an unaddressed review comment from when the feature was
first merged.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: fix and tweak west boards

- This script didn't get fixed when cmake.py was renamed zcmake, so it
  won't run; fix that.
- Change the default format string to '{name}' to keep things simple
- Flake8 lint

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
Rebased and made a few improvements.

@carlescufi I tacked a patch on at the beginning with a fix for west boards, as well as changing the default format to `{name}` and doing the flake8 linting we discussed.

mbolivar commented:
> Rebased and made a few improvements (ready for review).

Force pushes since this comment just had the effect of making `--board` at the command line take precedence over `BOARD` in the environment, making west build match CMake's own behavior (`-DBOARD` beats the environment).

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -143,6 +144,10 @@ def do_add_parser(self, parser_adder):
                             dest='dry_run', action='store_true',
                             help='''Just print the build commands which would
                             have run, without actually running them.''')
+        parser.add_argument('-o', '--build-opt', default=[], action='append',
```
`-o` is a bit weird but I really don't have a better suggestion

mbolivar replied:
Yeah, I struggled with this one.

carlescufi commented on doc/guides/west/build-flash-debug.rst:
````diff
@@ -60,7 +60,7 @@ exactly the same name you would supply to CMake if you were to invoke it with:
 .. tip::
 
-   You can use the :ref:```west boards`` <west-boards>` command to list all
+   You can use the :ref:`west boards <west-boards>` command to list all
````
this was fixed upstream as well, remove this commit

mbolivar replied:
Done

mbolivar commented on scripts/west_commands/boards.py:
```diff
@@ -10,7 +10,7 @@
 from west import log
 from west.commands import WestCommand
-from cmake import run_cmake
+from zcmake import run_cmake
```
Done

mbolivar commented:
Rebased and addressed @carlescufi's comments.

nashif commented:
how do I get verbose output from make/ninja? west build -v still does not work here.

mbolivar commented:
> how do I get verbose output from make/ninja? west build -v still does not work here.

By adding `-o=-v` (ninja) or `-o=VERBOSE=1` (make) for now, as discussed in the west issue. I'm going to try something out to try to special-case add these flags for those generators when `west -v build` is run in a follow-up.

nashif commented:
why is it better to do this:

`west build --build-opt "VERBOSE=1" -- -G'Unix Makefiles'`

if I can do this for both make and ninja:

`west build -b qemu_x86  samples/synchronization/  -- -DCMAKE_VERBOSE_MAKEFILE=ON`

nashif commented:
> By adding `-o=-v` (ninja) or `-o=VERBOSE=1` (make) for now, as discussed in the west issue. I'm going to try something out to try to special-case add these flags for those generators when `west -v build` is run in a follow-up.

great :)

mbolivar commented:
> why is it better to do this:

It's neither better nor worse. The behavior is different in each of your examples.

> west build --build-opt "VERBOSE=1" -- -G'Unix Makefiles'

This turns on verbose builds for just this build by passing `VERBOSE=1` to make. If you `west build` again, the build output is not verbose, because that argument will be missing.

> if I can do this for both make and ninja:
> west build -b qemu_x86 samples/synchronization/ -- -DCMAKE_VERBOSE_MAKEFILE=ON

This turns on verbose builds no matter what. If you `west build` again, the build output will be verbose, because the build system is configured to always be verbose.

dbkinder commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -27,88 +27,175 @@ directly delegate to West.
+This command attempts to "do what you mean" when run from a Zephyr application
+source or build directory:
+
+- When running ``west build`` in an existing build directory, the board, source
```
Better grammar would be:
```
When you run ``west build`` ...
```
It's like the sentence, "While walking down the street, the trees were waving in the wind."  where it's as if the trees were walking down the street.

dbkinder commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -27,88 +27,175 @@ directly delegate to West.
+Basics
+======
+
+The easiest way to use it is to go to an application's root directory (i.e. the
```
change ``it`` to ``west``

mbolivar replied:
Will change it to `west build`

dbkinder commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -27,88 +27,175 @@ directly delegate to West.
+======
+
+The easiest way to use it is to go to an application's root directory (i.e. the
+folder containing its :file:`CMakeLists.txt`) and then run::
```
change ``its`` to ``the application's``

dbkinder commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -27,88 +27,175 @@ directly delegate to West.
-  parameters) without having to specify the board again. If you're unsure
-  whether ``-b`` is required, just try leaving it out. West will print an
-  error if the option is required and was not given.
+You don't need to give ``-b`` if you've already got an existing build
```
change
```
give ``-b``
```
to
```
use the ``-b <BOARD>`` parameter

mbolivar replied:
I think "option" would be better than "parameter" in this case. Parameters generally are used when referring to functions rather than command line programs.

dbkinder reviewed (changes requested):
a few grammar tweaks...

mbolivar commented:
@dbkinder thanks for review! Comments addressed.

dbkinder reviewed (approved):
docs LGTM, thanks!
